### ISSUE TITLE
Add Video Prediction models

### ISSUE THREAD

**Issue by jacobbieker** [openclimatefix/satflow#4]
Various ones to try include https://github.com/thuml/predrnn-pytorch and ConvLSTMs, and potentially GANs as shown in https://arxiv.org/abs/2104.00954

**Comment by jacobbieker** [comment]
MetNet's work with axial self-attention did really well with precipitation forecasting of up to 8 hours later, and use axial self-attention, which there is a PyTorch implementation here: https://github.com/lucidrains/axial-attention

**Comment by jacobbieker** [comment]
The 12 spectral bands from SEVRI are not all the same resolution, with the non-HRV being 3km resolution, and HRV being 1km, so need to figure out best way to resample that or potentially just put the non-HRV values at 9 pixels in the HRV image, easily enough.

**Comment by JackKelly** [comment]
Yeah, in the Zarr file that I've been using, all 12 SEVIRI channels have been down-sampled to something like 4 km resolution.

We have a bunch of code for spatially re-projecting the raw SEVIRI data, and converting to Zarr, if that's useful.

**Comment by jacobbieker** [comment]
Yeah, I've been looking at that code and messing around with it currently, just working out the best way to integrate all the data together

**Comment by JackKelly** [comment]
Cool beans. I think the most recent reprojection code is in the [satip](https://github.com/openclimatefix/Satip) project.

I'd be tempted to start simple, and start with just one (or a small number) of satellite channels. (And perhaps ignore the High Res Visible channel to start with.  I've heard some people suggest that HRV might be one of the less informative channels when it comes to nowcasting clouds, and estimating irradiance at the Earth's surface!  But I could be wrong!)  That said, I can totally see the attraction of designing a data pipeline from the ground-up that can handle multiple, diverse datasets; so go for it if you want to start there!

**Comment by jacobbieker** [comment]
After the chat today, seems like a good one to try first would be UNET, MetNet, then something like DVD-GAN. Ideally, UNET should get fairly decent results.

**Comment by jacobbieker** [comment]
According to https://www.climatechange.ai/papers/icml2021/54 GANs show promise for predicting future frames from current ones, the code isn't available, but they seem to use a Recurrent Unet + cGAN in their slide.

**Comment by jacobbieker** [comment]
Pix2Pix does okay trying to predict the 12 satellite channels for the next half hour from the past 20min of data. I tried just half an hour in the future as it was easier to visualize and see what is happening, but plan on going up to 4 hours in the future soonish. This implementation is simple in that it takes a single input image, where the timeseries of images is just stacked along the channel dimension as input, and outputs a multi-channel output image in the same format, so each 12 channels is a single timestep. 

These are the generated images after 9 epochs of 2000*16 input image stacks. 
![generated_image](https://user-images.githubusercontent.com/7170359/126073111-31fe3a82-e15e-4cd0-a50b-eb7f871e4671.png)

And this is the ground truth for that same timestep
![target_image](https://user-images.githubusercontent.com/7170359/126073125-bc3b5c9e-a667-4e84-a664-79d15ef1e459.png)

As can be seen, Pix2Pix does get some of the structure present, but is very grainy looking, and fails on the last 2 channels. It does a bit better with this one:

Generated:
![generated_2](https://user-images.githubusercontent.com/7170359/126073403-03f3f248-35b5-4afb-b4fb-d1eb65d31591.png)

Target:
![target_2](https://user-images.githubusercontent.com/7170359/126073409-b9592310-89b4-4bab-83dc-9d0e4e005026.png)

**Comment by jacobbieker** [comment]
Another one is https://github.com/FrontierDevelopmentLab/PyRain which does global rain forecasting, mostly seems to be just ConvLSTM, but could have some good stuff on data preparation, etc. Or this one, which the code doesn't seem to be public, but they use a GAN for thunderstorm prediction http://spaceml.org/repo/project/60002402f5647f00129f7287

**Comment by jacobbieker** [comment]
https://github.com/CNALeon007/TrajGRU/blob/master/TrajGRU.py Trajectory GRU is also interesting, built specifically for precipitation forecasting, it could improve the tracking of clouds through time.

**Comment by jacobbieker** [comment]
Perciever IO now is also available for predicting dense predictions: https://github.com/deepmind/deepmind-research/tree/master/perceiver

**Comment by jacobbieker** [comment]
The perceiver IO is now also implemented in PyTorch here: https://github.com/lucidrains/perceiver-pytorch, https://github.com/openclimatefix/perceiver-pytorch/issues/1

**Comment by jacobbieker** [comment]
[PyTorch Video](https://pytorchvideo.readthedocs.io/en/latest/models.html) has some neat built in ones, and it might also be helpful to try out https://github.com/facebookresearch/AVT and https://github.com/facebookresearch/SlowFast
